vscode/src/debugger/session.ts

```diff
@@ -190,64 +209,90 @@ export class QscDebugSession extends LoggingDebugSession {
+      await this.continue();
     }
-    log.trace(`sending launchRequest response`);
-    this.sendResponse(response);
   }
 
-  private async runWithDebugging(
-    _args: ILaunchRequestArguments
-  ): Promise<void> {
-    const bps = this.getBreakpointIds();
-    this.run(bps);
-  }
-
-  private async run(bps: number[]): Promise<void> {
-    const eventTarget = createDebugConsoleEventTarget();
-    await this.debugService.evalContinue(bps, eventTarget).then(
-      (result) => {
-        if (result) {
-          log.trace(`raising breakpoint event`);
+  private async eval_step(step: () => Promise<IStructStepResult>) {
+    await step().then(
+      async (result) => {
+        if (result.id == StepResultId.BreakpointHit) {
           const evt = new StoppedEvent(
             "breakpoint",
             QscDebugSession.threadID
           ) as DebugProtocol.StoppedEvent;
-          evt.body.hitBreakpointIds = [result];
+          evt.body.hitBreakpointIds = [result.value];
+          log.trace(`raising breakpoint event`);
           this.sendEvent(evt);
+        } else if (result.id == StepResultId.Return) {
+          await this.endSession(`ending session`, 0);
         } else {
-          this.endSession(`ending session`);
+          log.trace(`step result: ${result.id} ${result.value}`);
+          this.sendEvent(new StoppedEvent("step", QscDebugSession.threadID));
         }
       },
       (error) => {
-        log.info(`ending session due to error: ${error}`);
-        vscode.debug.activeDebugConsole.appendLine("");
-        vscode.debug.activeDebugConsole.appendLine(SimulationCompleted);
-        this.sendEvent(new TerminatedEvent());
-        this.sendEvent(new ExitedEvent(0));
+        this.endSession(`ending session due to error: ${error}`, 1);
       }
     );
   }
 
-  private endSession(message: string) {
+  private async continue(): Promise<void> {
+    const bps = this.getBreakpointIds();
+    await this.eval_step(
+      async () => await this.debugService.evalContinue(bps, this.eventTarget)
+    );
+  }
+
+  private async next(): Promise<void> {
+    const bps = this.getBreakpointIds();
+    await this.eval_step(
+      async () => await this.debugService.evalNext(bps, this.eventTarget)
+    );
+  }
+
+  private async stepIn(): Promise<void> {
+    const bps = this.getBreakpointIds();
+    await this.eval_step(
+      async () => await this.debugService.evalStepIn(bps, this.eventTarget)
+    );
+  }
+
+  private async stepOut(): Promise<void> {
+    const bps = this.getBreakpointIds();
+    await this.eval_step(
+      async () => await this.debugService.evalStepOut(bps, this.eventTarget)
+    );
+  }
+
+  private async endSession(message: string, exitCode: number): Promise<void> {
     log.trace(message);
-    vscode.debug.activeDebugConsole.appendLine("");
-    vscode.debug.activeDebugConsole.appendLine(SimulationCompleted);
+    this.writeToStdOut("");
+    this.writeToStdOut(SimulationCompleted);
```

Comment:
May be me being picky, but I wouldn't expect anything printed to stdout that isn't coming from the program itself (like `Message()` output.) Maybe this could go to the debug console?